File: modules/core/core-api/src/main/java/com/enonic/xp/security/AuthConfig.java
package com.enonic.xp.security;


import java.util.Objects;

import com.google.common.annotations.Beta;
import com.google.common.base.Preconditions;

import com.enonic.xp.app.ApplicationKey;
import com.enonic.xp.data.PropertyTree;

@Beta
public final class AuthConfig
{
    private final ApplicationKey applicationKey;

    private final PropertyTree config;

    private AuthConfig( final Builder builder )
    {
        Preconditions.checkNotNull( builder.applicationKey, "applicationKey cannot be null" );
        Preconditions.checkNotNull( builder.config, "config cannot be null" );
        this.applicationKey = builder.applicationKey;
        this.config = builder.config;
    }

    public ApplicationKey getApplicationKey()
    {
        return applicationKey;
    }

    public PropertyTree getConfig()
    {
        return config;
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }

        final AuthConfig that = (AuthConfig) o;

        return Objects.equals( this.applicationKey, that.applicationKey ) && Objects.equals( this.config, that.config );
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( applicationKey, config );
    }

    public static Builder create()
    {
        return new Builder();
    }

    public static class Builder
    {
        private ApplicationKey applicationKey;

        private PropertyTree config;

        public Builder applicationKey( ApplicationKey value )
        {
            this.applicationKey = value;
            return this;
        }

        public Builder config( PropertyTree value )
        {
            this.config = value;
            return this;
        }

        public AuthConfig build()
        {
            return new AuthConfig( this );
        }
    }
}

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/security/AuthConfig.ts
module api.security {
    export class AuthConfig implements api.Equitable {
        private applicationKey: api.application.ApplicationKey;
        private config: api.data.PropertyTree;

        constructor(builder: AuthConfigBuilder) {
            this.applicationKey = builder.applicationKey;
            this.config = builder.config;
        }

        getApplicationKey(): api.application.ApplicationKey {
            return this.applicationKey;
        }

        getConfig(): api.data.PropertyTree {
            return this.config;
        }

        equals(o: api.Equitable): boolean {
            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, AuthConfig)) {
                return false;
            }

            var other = <AuthConfig> o;

            return this.applicationKey.equals(other.applicationKey) &&
                   this.config.equals(other.config);
        }

        toJson(): AuthConfigJson {
            return {
                "applicationKey": this.applicationKey.toString(),
                "config": this.config.toJson()
            };
        }

        clone(): AuthConfig {
            return AuthConfig.create().
                setApplicationKey(this.applicationKey).
                setConfig(this.config.copy()).
                build();
        }

        static create(): AuthConfigBuilder {
            return new AuthConfigBuilder();
        }

        static fromJson(json: AuthConfigJson): AuthConfig {
            return new AuthConfigBuilder().fromJson(json).build();
        }

    }

    export class AuthConfigBuilder {
        applicationKey: api.application.ApplicationKey;
        config: api.data.PropertyTree;

        setApplicationKey(applicationKey: api.application.ApplicationKey): AuthConfigBuilder {
            this.applicationKey = applicationKey;
            return this;
        }

        setConfig(config: api.data.PropertyTree): AuthConfigBuilder {
            this.config = config;
            return this;
        }


        fromJson(json: api.security.AuthConfigJson): AuthConfigBuilder {
            this.applicationKey = api.application.ApplicationKey.fromString(json.applicationKey);
            this.config = json.config != null ? api.data.PropertyTree.fromJson(json.config) : null;
            return this;
        }

        build(): AuthConfig {
            return new AuthConfig(this);
        }
    }
}